src/Orleans/Core/InterceptedMethodInvokerCache.cs

```diff
@@ -30,18 +48,44 @@ public InterceptedMethodInvoker GetInterceptedMethodInvoker(Type implementationT
             // Create a mapping between the interface and the implementation.
             return implementation.GetOrAdd(
                 interfaceId,
-                _ =>
-                    new InterceptedMethodInvoker(invoker,
-                        GetInterfaceToImplementationMap(interfaceId, implementationType)));
+                _ => CreateInterceptedMethodInvoker(implementationType, interfaceId, invoker));
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="InterceptedMethodInvoker"/> for the provided values.
+        /// </summary>
+        /// <param name="implementationType">The grain type.</param>
+        /// <param name="interfaceId">The interface id.</param>
+        /// <param name="invoker">The invoker.</param>
+        /// <returns>A new <see cref="InterceptedMethodInvoker"/> for the provided values.</returns>
+        private static InterceptedMethodInvoker CreateInterceptedMethodInvoker(
+            Type implementationType,
+            int interfaceId,
+            IGrainMethodInvoker invoker)
+        {
+            // If a grain extension is being invoked, the implementation map must match the methods on the extension
+            // and not the grain implementation.
+            var extensionMap = invoker as IGrainExtensionMap;
```

Comment:
This is the functional core of the fix: if the invoker is the extension invoker, then we retrieve the extension class and get the implementation map for that class instead of the grain class.